src/main/java/org/snomed/snowstorm/ecl/validation/ECLEdgeCaseHandlerService.java

```diff
@@ -0,0 +1,131 @@
+package org.snomed.snowstorm.ecl.validation;
+
+import com.google.common.base.Functions;
+import io.kaicode.elasticvc.api.CommitListener;
+import io.kaicode.elasticvc.api.VersionControlHelper;
+import io.kaicode.elasticvc.domain.Commit;
+import org.snomed.langauges.ecl.ECLQueryBuilder;
+import org.snomed.langauges.ecl.domain.refinement.SubAttributeSet;
+import org.snomed.snowstorm.ecl.ECLQueryService;
+import org.snomed.snowstorm.ecl.domain.expressionconstraint.SExpressionConstraint;
+import org.snomed.snowstorm.ecl.domain.expressionconstraint.SRefinedExpressionConstraint;
+import org.snomed.snowstorm.ecl.domain.expressionconstraint.SSubExpressionConstraint;
+import org.snomed.snowstorm.ecl.domain.refinement.*;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.domain.PageRequest;
+import org.springframework.stereotype.Service;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+import static org.snomed.snowstorm.core.data.domain.Concepts.CONCEPT_MODEL_DATA_ATTRIBUTE;
+import static org.snomed.snowstorm.ecl.domain.expressionconstraint.SExpressionConstraintHelper.MISSING;
+
+@Service
+public class ECLEdgeCaseHandlerService implements CommitListener {
+
+	private static final String RETURN_ALL_CONCRETE_ATTRIBUTES_ECL_QUERY = "< " + CONCEPT_MODEL_DATA_ATTRIBUTE;
+
+	private static final List<String> CACHED_CONCRETE_CONCEPT_IDS = new ArrayList<>();
+
+	@Autowired
+	private ECLQueryBuilder eclQueryBuilder;
```

Comment:
It would be better separation of concerns if this class took SExpressionConstraint as input and output. It does not need to be involved in calling the ECL builder.